### ISSUE TITLE
Unexpected failure when terminating trio components during shutdown

### ISSUE THREAD

**Issue by gsalgado** [ethereum/trinity#1751]
### What is wrong?

On python3.8, I'm seeing this quite frequently

```
   DEBUG  2020-05-27 18:41:46,265  asyncio_run_in_process  Got CancelledError while running subprocess pid=105840.  Sending SIGINT.
...
   DEBUG  2020-05-27 18:41:46,733  asyncio_run_in_process  process Process[<bound method TrioIsolatedComponent._do_run of <trinity.components.builtin.upnp.component.UpnpComponent object at 0x7fea90ccc5e0>>] (pid=105840) finished: returncode=0
...
   DEBUG  2020-05-27 18:41:48,266  asyncio_run_in_process  Timed out waiting for pid=105840 to exit after SIGINT, sending SIGTERM
[1m[31m   ERROR  2020-05-27 18:41:48,267  asyncio_run_in_process  Unexpected error when terminating child; pid=105840
Traceback (most recent call last):
  File "/home/salgado/virtualenvs/trinity/lib/python3.8/site-packages/asyncio_run_in_process/run_in_process.py", line 274, in _open_in_process
    yield proc
  File "/home/salgado/src/trinity/trinity/extensibility/trio.py", line 38, in run
    await proc.wait_result_or_raise()
  File "/home/salgado/virtualenvs/trinity/lib/python3.8/site-packages/asyncio_run_in_process/process.py", line 230, in wait_result_or_raise
    await self.wait_returncode()
  File "/home/salgado/virtualenvs/trinity/lib/python3.8/site-packages/asyncio_run_in_process/process.py", line 174, in wait_returncode
    await self._has_returncode.wait()
  File "/usr/lib/python3.8/asyncio/locks.py", line 309, in wait
    await fut
asyncio.exceptions.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/salgado/virtualenvs/trinity/lib/python3.8/site-packages/asyncio_run_in_process/run_in_process.py", line 306, in _open_in_process
    await asyncio.wait_for(
  File "/usr/lib/python3.8/asyncio/tasks.py", line 490, in wait_for
    raise exceptions.TimeoutError()
asyncio.exceptions.TimeoutError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/salgado/virtualenvs/trinity/lib/python3.8/site-packages/asyncio_run_in_process/run_in_process.py", line 313, in _open_in_process
    proc.terminate()
  File "/home/salgado/virtualenvs/trinity/lib/python3.8/site-packages/asyncio_run_in_process/process.py", line 266, in terminate
    self.send_signal(signal.SIGTERM)
  File "/home/salgado/virtualenvs/trinity/lib/python3.8/site-packages/asyncio_run_in_process/process.py", line 272, in send_signal
    os.kill(self.pid, signum.value)
ProcessLookupError: [Errno 3] No such process[0m
``` 

It looks like `asyncio-run-in-process` believes the process is still running when it no longer is, and then we get an error when it tries to forcibly terminate the process

**Comment by gsalgado** [comment]
This seems to have a similar root cause to that of #1752, but here `_monitor_state()` [loops through **all** expected states, then hangs waiting for the sub proc's return value, on line 158](https://github.com/ethereum/asyncio-run-in-process/blob/959f6116f4b06ee6d2b3a3e8abaa7ea33e957aca/asyncio_run_in_process/run_in_process.py#L108-L158) and then it never [sets the process' return_value](https://github.com/ethereum/asyncio-run-in-process/blob/959f6116f4b06ee6d2b3a3e8abaa7ea33e957aca/asyncio_run_in_process/run_in_process.py#L163], so the [`proc.wait()` in _open_in_process()](https://github.com/ethereum/asyncio-run-in-process/blob/959f6116f4b06ee6d2b3a3e8abaa7ea33e957aca/asyncio_run_in_process/run_in_process.py#L307) times out and we try to `terminate()` a process that no longer exists

**Comment by gsalgado** [comment]
Again, it's hanging on that `loop.run_in_executor()` call because the default executor has no available worker threads (https://github.com/ethereum/asyncio-run-in-process/pull/26), as trinity starts more processes than the max number of worker threads (8) in the default executor for my machine

**Comment by gsalgado** [comment]
Fixed in #1801
